Refactor 654 zonal mean xy

chengzhuzhang opened the pull request:
## Description

<!--
  Please include a summary of the change and which issue is fixed.
  Please also include relevant motivation and context.
  List any dependencies that are required for this change.
-->

- Closes #654  to refactor zonal mean xy to use xcdat/xarray

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] My changes generate no new warnings: there is warning from `xgcm`
- [x] Any dependent changes have been merged and published in downstream modules

If applicable:

- [ ] New and existing unit tests pass with my changes (locally and CI/CD build)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have noted that this is a breaking change for a major release (fix or feature that would cause existing functionality to not work as expected)

chengzhuzhang commented:
The components to refactor include the driver and the plotter.
File 1, **zonal_mean_xy_driver.py**

- [ ] refactor `regrid_to_lower_res_1d` function that  regrids 1-D transient variable toward lower resolution of two variables . Question can I reuse or extend `align_grids_to_lower_res`.

File 2, **zonal_mean_xy_plot.py**

- [ ] use xarray object instead of cdms transient variable. No major function to refactor.

chengzhuzhang commented:
With the new commit, there are two problems:
1. an error:
```
 2023-12-01 16:17:54,212 [ERROR]: core_parameter.py(_run_diag:326) >> Error in e3sm_diags.driver.zonal_mean_xy_driver
Traceback (most recent call last):
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/parameter/core_parameter.py", line 323, in _run_diag
    single_result = module.run_diag(self)
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/driver/zonal_mean_xy_driver.py", line 88, in run_diag
    _run_diags_3d(
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/driver/zonal_mean_xy_driver.py", line 200, in _run_diags_3d
    ds_test_rg = regrid_z_axis_to_plevs(ds_test, var_key, parameter.plevs)
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/driver/utils/regrid.py", line 384, in regrid_z_axis_to_plevs
    ds_plevs = _hybrid_to_plevs(ds, var_key, plevs)
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/driver/utils/regrid.py", line 443, in _hybrid_to_plevs
    result = ds.regridder.vertical(
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xcdat/regridder/accessor.py", line 392, in vertical
    input_grid = _get_input_grid(
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xcdat/regridder/accessor.py", line 445, in _get_input_grid
    grid = input_grid.regridder.grid
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xcdat/regridder/accessor.py", line 91, in grid
    data, bnds = self._get_axis_data(axis)
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xcdat/regridder/accessor.py", line 111, in _get_axis_data
    _validate_grid_has_single_axis_dim(name, coord_var)
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xcdat/regridder/grid.py", line 744, in _validate_grid_has_single_axis_dim
    raise ValueError(
ValueError: Multiple 'X' axis dims were found in this dataset, ['lon', 'slon']. Please drop the unused dimension(s) before performing grid operations.
```
2. pre-commiting hooks `isort` and `black` conflicts. And updating .pre-committing yaml didn't help.

tomvothecoder commented on e3sm_diags/driver/zonal_mean_xy_driver.py:
```diff
@@ -1,262 +1,316 @@
-from e3sm_diags.plot import plot
+
+# TODO
+# from e3sm_diags.driver.utils.io import _save_data_metrics_and_plots
```
This commented line was higher in the import list which caused `isort` to throw an error about a blank line in between imports. Moving it down here fixes that issue for now.

I removed the `.pre-commit-config.yaml` file update for `isort` to use the `black` profile since it isn't needed.

tomvothecoder commented on e3sm_diags/driver/zonal_mean_xy_driver.py:
```diff
@@ -1,262 +1,316 @@
+    """
+    for region in regions:
+        parameter._set_param_output_attrs(var_key, season, region, ref_name, ilev=None)
+        # TODO: Fix type annotation
```
I added a few TODO: statements to fix the `mypy` type annotation issues.

Most of them come up because the return type for `_calculate_zonal_mean()`  hasn't been updated.

tomvothecoder reviewed (commented):
Hey @chengzhuzhang, I addressed your recent issues in this [comment](https://github.com/E3SM-Project/e3sm_diags/pull/752#issuecomment-1836953913).

I also rebased this branch on the latest version of `cdat-migration-fy24`, which I squashed down into less commits.

chengzhuzhang commented:
@tomvothecoder thank you for fixing this branch!

For the excessive `slat`, `slon`, it turned out to be they are included in the model climatology files, the relevant texts in the header as shown below:
```
        double slat(slat) ;
                slat:long_name = "Latitude for staggered FV grid" ;
                slat:units = "degrees_north" ;
        double w_stag(slat) ;
                w_stag:long_name = "Latitude weights for staggered FV grid" ;
        double slon(slon) ;
                slon:long_name = "Longitude for staggered FV grid" ;
                slon:units = "degrees_east" ;
```
I think I will try dropping the `slat` and `slon` dims, as a workaround for now.

chengzhuzhang commented:
> Hey @chengzhuzhang, I addressed your recent issues in this [comment](https://github.com/E3SM-Project/e3sm_diags/pull/752#issuecomment-1836953913).
> 
> I also rebased this branch on the latest version of `cdat-migration-fy24`, which I squashed down into less commits. If you have any un-committed changes, can you stash then, check out the latest version of this branch, then pop them back on top? Thanks

yep, I have checked out the latest version. Thanks for the heads-up!

tomvothecoder commented on examples/test_refactor/ex5.py:
```diff
@@ -0,0 +1,38 @@
+import os
```
```suggestion
import os
import sys
```

tomvothecoder commented on examples/test_refactor/ex5.py:
```diff
@@ -0,0 +1,38 @@
+# For running with multiprocessing.
+# param.multiprocessing = True
+# param.num_workers = 32
+
+runner.sets_to_run = ["zonal_mean_xy"]
+runner.run_diags([param])
```
```suggestion
# Use the specified `.cfg` file for debugging
CFG_PATH = "examples/test_refactor/diags.cfg"
sys.argv.extend(["-d", CFG_PATH])

runner.sets_to_run = ["zonal_mean_xy"]
runner.run_diags([param])
```

tomvothecoder reviewed (commented):
Just some updates to your test script to pass `diags.cfg` using `sys.argv.extend`. This allows you to perform Python debugging with parameters defined in `diags.cfg`.

chengzhuzhang commented:
@tomvothecoder I made plotting work, but will try consolidate the codes a little better. 
But it seems my pre-commit isort still has errorsthat I can't sort out..

chengzhuzhang commented:
Other than `mypy` complaints. There are still several problems causing some variables fails to generate plots. [The log for a full run](https://portal.nersc.gov/cfs/e3sm/chengzhu/test_e3sm_refactor/zonal_mean_xy_model_to_obs_ANN/prov/e3sm_diags_run.log).
The result can be found [here](https://portal.nersc.gov/cfs/e3sm/chengzhu/test_e3sm_refactor/zonal_mean_xy_model_to_obs_ANN/viewer/zonal_mean_xy/index.html)

tomvothecoder commented:
> Other than `mypy` complaints. There are still several problems causing some variables fails to generate plots. [The log for a full run](https://portal.nersc.gov/cfs/e3sm/chengzhu/test_e3sm_refactor/zonal_mean_xy_model_to_obs_ANN/prov/e3sm_diags_run.log). The result can be found [here](https://portal.nersc.gov/cfs/e3sm/chengzhu/test_e3sm_refactor/zonal_mean_xy_model_to_obs_ANN/viewer/zonal_mean_xy/index.html)

I listed the errors from the log below:

### Issue 1
```python
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/driver/zonal_mean_xy_driver.py", line 322, in regrid_to_lower_res_1d
    if len(ds_test_1d.lat) > len(ds_ref_1d.lat):
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/core/common.py", line 278, in __getattr__
    raise AttributeError(
AttributeError: 'DataArray' object has no attribute 'lat'
```
  * The name of the latitude coordinates is not lat on the DataArray. You should use `xc.get_dim_keys(DataArray, axis="Y")` to get the name dynamically.

### Issue 2
```python
packages/e3sm_diags/plot/zonal_mean_xy_plot.py:115: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`). Consider using `matplotlib.pyplot.close()`.
  fig = plt.figure(figsize=parameter.figsize, dpi=parameter.dpi)
```
  * The warning message is self-explanatory.

### Issue 3
```python
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/plot/zonal_mean_xy_plot.py", line 130, in plot
    ax1.set_ylabel(da_test.long_name + " (" + da_test.units + ")")
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/core/common.py", line 278, in __getattr__
    raise AttributeError(
AttributeError: 'DataArray' object has no attribute 'long_name'
```
* The `long_name` attribute looks like it is dropped at some point. Maybe when calculating zonal mean?

### Issue 4
```python
Traceback (most recent call last):
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/coding/times.py", line 213, in _decode_cf_datetime_dtype
    result = decode_cf_datetime(example_value, units, calendar, use_cftime)
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/coding/times.py", line 321, in decode_cf_datetime
    dates = _decode_datetime_with_cftime(
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/coding/times.py", line 237, in _decode_datetime_with_cftime
    cftime.num2date(num_dates, units, calendar, only_use_cftime_datetimes=True)
  File "src/cftime/_cftime.pyx", line 580, in cftime._cftime.num2date
  File "src/cftime/_cftime.pyx", line 98, in cftime._cftime._dateparse
ValueError: 'months since' units only allowed for '360_day' calendar
```
* Self-explanatory. The `cftime` library can't decode `months since" units with other calendar types. `xc.open_dataset()`/`xc.open_mf_dataset()` can decode non-CF time units. If one of these functions is used, maybe remove `use_cftime=True` to avoid using  Xarray's `cftime` decoder.

### Issue 5
```python
Traceback (most recent call last):
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/conventions.py", line 432, in decode_cf_variables
    new_vars[k] = decode_cf_variable(
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/conventions.py", line 283, in decode_cf_variable
    var = times.CFDatetimeCoder(use_cftime=use_cftime).decode(var, name=name)
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/coding/times.py", line 832, in decode
    dtype = _decode_cf_datetime_dtype(data, units, calendar, self.use_cftime)
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/coding/times.py", line 223, in _decode_cf_datetime_dtype
    raise ValueError(msg)
ValueError: unable to decode time units 'months since 1983-06' with 'the default calendar'. Try opening your dataset with decode_times=False or installing cftime if it is not installed.
```
* Self-explanatory. The `cftime` library can't decode `months since" units with other calendar types. `xc.open_dataset()`/`xc.open_mf_dataset()` can decode non-CF time units. If one of these functions is used, maybe remove `use_cftime=True` to avoid using  Xarray's `cftime` decoder.

### Issue 6
```python
 File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/driver/utils/dataset_xr.py", line 254, in _get_global_attr_from_climo_dataset
    ds = xr.open_dataset(filepath)
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/backends/api.py", line 572, in open_dataset
    backend_ds = backend.open_dataset(
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/backends/netCDF4_.py", line 607, in open_dataset
    ds = store_entrypoint.open_dataset(
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/backends/store.py", line 58, in open_dataset
    ds = Dataset(vars, attrs=attrs)
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/core/dataset.py", line 697, in __init__
    variables, coord_names, dims, indexes, _ = merge_data_and_coords(
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/core/dataset.py", line 426, in merge_data_and_coords
    return merge_core(
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/core/merge.py", line 724, in merge_core
    dims = calculate_dimensions(variables)
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/core/variable.py", line 3007, in calculate_dimensions
    raise ValueError(
ValueError: dimension 'time' already exists as a scalar variable
```
* _Not 100% what is happening here when you open up the dataset._

### Issue 7
```python
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/derivations/utils.py", line 175, in cosp_bin_sum
    prs: FileAxis = cld.getAxis(0)
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/core/common.py", line 278, in __getattr__
    raise AttributeError(
AttributeError: 'DataArray' object has no attribute 'getAxis'
```
* The `cosp_bin_sum()` and `cosp_histogram_standardize()` functions are refactored to use Xarray/xCDAT in this PR: https://github.com/E3SM-Project/e3sm_diags/pull/748/files#diff-2fd1bd98452189f43f530975ba12afc289337494f779cb5a2f97144d30820f22
* We need to merge this PR first before this issue can be fixed.

### Issue 8
```python
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/derivations/formulas.py", line 81, in pminuse_convert_units
    var.attrs["units"] == "kg/m2/s"
KeyError: 'units'
```
* The `units` attribute looks like it is dropped at some point. Maybe when calculating zonal mean?

tomvothecoder commented on e3sm_diags/driver/utils/io.py:
```diff
@@ -68,13 +70,17 @@ def _save_data_metrics_and_plots(
-    plot_func(
+    # Get the function arguments and pass to the set's plotting function.
+    args = [
         parameter,
         ds_test[var_key],
         ds_ref[var_key] if ds_ref is not None else None,
         ds_diff[var_key] if ds_diff is not None else None,
-        metrics_dict,
-    )
+    ]
+    if metrics_dict is not None:
+        args = args + [metrics_dict]
+
+    plot_func(*args)
```
Updated `_save_data_metrics_and_plots()` to make `metrics_dict` optional for the `plot_func` (the set's plot function).

I pulled this from my PR #748.

tomvothecoder commented on e3sm_diags/driver/zonal_mean_xy_driver.py:
```diff
@@ -45,9 +47,15 @@ def run_diag(parameter: CoreParameter) -> CoreParameter:
+            raise RuntimeError(
+                f"Region ({region}) is not supported. Only global region is currently "
+                "supported for the zonal_mean_xy set."
+            )
```
I moved this check out of `_calculate_zonal_mean()` to the top-level of `run_diags()`.

tomvothecoder commented on e3sm_diags/driver/zonal_mean_xy_driver.py:
```diff
@@ -201,129 +205,83 @@ def _run_diags_3d(
+    """Calculate zonal mean metrics.
+
+    # TODO: Write unit tests for this function.
 
```
I removed args from `_calc_zonal_mean()` that aren't needed anymore.

tomvothecoder commented on e3sm_diags/driver/zonal_mean_xy_driver.py:
```diff
@@ -201,129 +205,83 @@ def _run_diags_3d(
-        ds_test_1d = zonal_avg(ds_test, var_key)
-        ds_ref_1d = None
 
-    return ds_test_1d, ds_ref_1d  # type: ignore
```
`_get_diff_of_zonal_means()` and the plotter function in `zonal_mean_xy_plot.py` always expect a reference dataset, so we can't `ds_ref_1d =  None` (mypy warnings). 

Instead, always take the spatial average of test and ref datasets.

tomvothecoder commented on e3sm_diags/driver/zonal_mean_xy_driver.py:
```diff
@@ -201,129 +205,83 @@ def _run_diags_3d(
-    results = ds_avg[var_key]
-
-    return results
+    with xr.set_options(keep_attrs=True):
```
This line might resolve a few of the missing attribute issues from your log ([comment](https://github.com/E3SM-Project/e3sm_diags/pull/752#issuecomment-1879174866)).

tomvothecoder commented on e3sm_diags/driver/zonal_mean_xy_driver.py:
```diff
@@ -201,129 +205,83 @@ def _run_diags_3d(
-        da_ref_1d_new = da_test_1d.copy(data=interpf(da_test_1d.lat))
-        da_diff_1d = da_test_1d - da_ref_1d_new
-    return da_diff_1d
+            return da_a.copy() - da_b_new
```
Updated names of arguments and variables.

tomvothecoder commented on e3sm_diags/driver/zonal_mean_xy_driver.py:
```diff
@@ -201,129 +205,83 @@ def _run_diags_3d(
-
-    return results
+    with xr.set_options(keep_attrs=True):
+        if len(da_a.lat) > len(da_b.lat):
```
FIXME: This line needs to be updated to dynamically get the latitude coordinates.
We can use `xc.get_dim_coords(da_a, axis="Y")` and `xc.get_dim_coords(da_b, axis="Y")`.

This might fix one of the issues in [comment](https://github.com/E3SM-Project/e3sm_diags/pull/752#issuecomment-1879174866)).

tomvothecoder commented on e3sm_diags/driver/zonal_mean_xy_driver.py:
```diff
@@ -201,129 +205,83 @@ def _run_diags_3d(
     """
-    logger.info(f"Selected region: {region}")
-    parameter.var_region = region
+    da_test_1d = ds_test.spatial.average(var_key, axis="X", weights="generate")[var_key]
+    da_ref_1d = ds_ref.spatial.average(var_key, axis="X", weights="generate")[var_key]
```
It's simpler to just call the xCDAT spatial averaging API here.

chengzhuzhang replied:
Just to clarify do you mean it is simpler not wrapping both in a function?

chengzhuzhang replied:
@tomvothecoder I"m about to wrapping up updating this PR. Not sure how best to address this.

tomvothecoder replied:
I replaced your  `zonal_mean()` function with these calls to xCDAT directly.

We can actually re-use `e3sm_diags.metrics.spatial_avg`, which was updated to accept an `axis` arg. I'll replace these calls when I do a review.

tomvothecoder commented on e3sm_diags/plot/zonal_mean_xy_plot.py:
```diff
@@ -36,67 +36,11 @@
     da_test: xr.DataArray,
-    da_ref: xr.DataArray | None,
-    da_diff: xr.DataArray | None,
-    metrics_dict,
```
* Updated type annotations for `da_ref` and `da_diff` to make them required. 
* Remove `metrics_dict` because it is unused here and no longer required

tomvothecoder commented on e3sm_diags/plot/zonal_mean_xy_plot.py:
```diff
@@ -169,3 +111,58 @@ def plot(
     _save_plot(fig, parameter)
+
+
+def _save_plot(fig: plt.figure, parameter: CoreParameter):
+    """Save the plot using the figure object and parameter configs.
+
+    This function creates the output filename to save the plot. It also
+    saves each individual subplot if the reference name is an empty string ("").
```
Moved this function to the bottom because it is a private function. Public functions should be at the top to make it easier to read the Python file and follow the order of function references (top-down approach).

forsyth2 replied:
@tomvothecoder @chengzhuzhang I'm confused as to why you're making your own `_save_plot` function here. Why wouldn't we just always use [e3sm_diags/plot/utils.py](https://github.com/E3SM-Project/e3sm_diags/blob/cdat-migration-fy24/e3sm_diags/plot/utils.py)'s `_save_plot` function?

chengzhuzhang replied:
@forsyth2 I made this design choice for zonal_mean_xy is because the the layout of the panels and plots are largely different than the 3 panel plots for lat-lon maps. i.g., For zonal_mean_xy, it contains two panels, with each a line plot. To reuse the _save_plot function in utils.py would introduce lots of conditional code blocks, which is not ideal.

tomvothecoder replied:
I actually did a diff of your version and the general `e3sm_diags.plot.utils._save_plot` and only found two differences involving setting the panel config and border_padding. I updated the general version with two new args for each config and removed your version.

Here is the commit with that change: e3sm_diags/plot/utils.py

tomvothecoder replied:
Screenshot of the diff:

![image](https://github.com/E3SM-Project/e3sm_diags/assets/25624127/e610ecdc-ed4a-487f-97fd-524209efb6bb)

chengzhuzhang replied:
Oh, I got it, thanks for the change @tomvothecoder . @forsyth2 good point. I misinterpreted the question, it is not ideal to have a similar _save_plot function!

tomvothecoder reviewed (commented):
Hi @chengzhuzhang, I did some code clean up code and fixed the mypy warnings (which were all definitely useful).

tomvothecoder commented on e3sm_diags/driver/zonal_mean_xy_driver.py:
```diff
@@ -1,262 +1,287 @@
+                "supported for the zonal_mean_xy set."
             )
 
-            # Special case, cdms didn't properly convert mask with fill value
-            # -999.0, filed issue with Denis.
-            if ref_name == "WARREN":
-                # This is cdms2 fix for bad mask, Denis' fix should fix this.
-                mv2 = MV2.masked_where(mv2 == -0.9, mv2)
-            # The following should be moved to a derived variable.
-            if ref_name == "AIRS":
-                # This is cdms2 fix for bad mask, Denis' fix should fix this.
-                mv2 = MV2.masked_where(mv2 > 1e20, mv2)
-            if ref_name == "WILLMOTT" or ref_name == "CLOUDSAT":
-                # This is cdms2 fix for bad mask, Denis' fix should fix this.
-                mv2 = MV2.masked_where(mv2 == -999.0, mv2)
-
-                # The following should be moved to a derived variable.
-                if var == "PRECT_LAND":
-                    days_season = {
-                        "ANN": 365,
-                        "DJF": 90,
-                        "MAM": 92,
-                        "JJA": 92,
-                        "SON": 91,
-                    }
-                    # mv1 = mv1 * days_season[season] * 0.1 # following AMWG
-                    # Approximate way to convert to seasonal cumulative
-                    # precipitation, need to have solution in derived variable,
-                    # unit convert from mm/day to cm.
-                    mv2 = (
-                        mv2 / days_season[season] / 0.1
-                    )  # Convert cm to mm/day instead.
-                    mv2.units = "mm/day"
```
I noticed this entire section is removed without any new Xarray code to replace it. Is this block of code no longer needed since they are "cdms2" fixes?

I want to make sure because this chunk of code is copy-pasted in `zonal_mean_2d_driver.py`, which I am refactoring in #774.

https://github.com/E3SM-Project/e3sm_diags/blob/3e49671dd732b0221baf3e164097a9895088bd66/e3sm_diags/driver/zonal_mean_2d_driver.py#L113-L142

chengzhuzhang replied:
Tom, I really appreciate the pointers you provided here and the fixes and updates to the PR. I'm resuming this work. 

For this question, this block of codes were useful at the very beginning of e3sm diags,( when its name was acme diags) for a set of legacy datasets inherited from AMWG package. They are no longer needed when we compile the new obs data inventory. They should have been removed earlier, and we can clean themn up during the migration.

tomvothecoder replied:
Of course! And thanks for confirming that we can deprecate this code.

This section is found in the polar driver as well. I'll leave a comment for @forsyth2 in #749 so that he is aware.

chengzhuzhang commented:
> > Other than `mypy` complaints. There are still several problems causing some variables fails to generate plots. [The log for a full run](https://portal.nersc.gov/cfs/e3sm/chengzhu/test_e3sm_refactor/zonal_mean_xy_model_to_obs_ANN/prov/e3sm_diags_run.log). The result can be found [here](https://portal.nersc.gov/cfs/e3sm/chengzhu/test_e3sm_refactor/zonal_mean_xy_model_to_obs_ANN/viewer/zonal_mean_xy/index.html)
> 
> I listed the errors from the log below:

> ### Issue 4
> 
> ```python
> Traceback (most recent call last):
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/coding/times.py", line 213, in _decode_cf_datetime_dtype
>     result = decode_cf_datetime(example_value, units, calendar, use_cftime)
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/coding/times.py", line 321, in decode_cf_datetime
>     dates = _decode_datetime_with_cftime(
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/coding/times.py", line 237, in _decode_datetime_with_cftime
>     cftime.num2date(num_dates, units, calendar, only_use_cftime_datetimes=True)
>   File "src/cftime/_cftime.pyx", line 580, in cftime._cftime.num2date
>   File "src/cftime/_cftime.pyx", line 98, in cftime._cftime._dateparse
> ValueError: 'months since' units only allowed for '360_day' calendar
> ```
> 
>     * Self-explanatory. The `cftime` library can't decode `"months since"` units with other calendar types. `xc.open_dataset()`/`xc.open_mf_dataset()` can decode non-CF time units. If one of these functions is used, maybe remove `use_cftime=True` to avoid using  Xarray's `cftime` decoder.
> 
> 
> ### Issue 5
> 
> ```python
> Traceback (most recent call last):
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/conventions.py", line 432, in decode_cf_variables
>     new_vars[k] = decode_cf_variable(
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/conventions.py", line 283, in decode_cf_variable
>     var = times.CFDatetimeCoder(use_cftime=use_cftime).decode(var, name=name)
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/coding/times.py", line 832, in decode
>     dtype = _decode_cf_datetime_dtype(data, units, calendar, self.use_cftime)
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/coding/times.py", line 223, in _decode_cf_datetime_dtype
>     raise ValueError(msg)
> ValueError: unable to decode time units 'months since 1983-06' with 'the default calendar'. Try opening your dataset with decode_times=False or installing cftime if it is not installed.
> ```
> 
>     * Self-explanatory. The `cftime` library can't decode `months since" units with other calendar types. `xc.open_dataset()`/`xc.open_mf_dataset()`can decode non-CF time units. If one of these functions is used, maybe remove`use_cftime=True`to avoid using  Xarray's`cftime` decoder.
> 
> 
> ### Issue 6
> 
> ```python
>  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/driver/utils/dataset_xr.py", line 254, in _get_global_attr_from_climo_dataset
>     ds = xr.open_dataset(filepath)
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/backends/api.py", line 572, in open_dataset
>     backend_ds = backend.open_dataset(
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/backends/netCDF4_.py", line 607, in open_dataset
>     ds = store_entrypoint.open_dataset(
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/backends/store.py", line 58, in open_dataset
>     ds = Dataset(vars, attrs=attrs)
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/core/dataset.py", line 697, in __init__
>     variables, coord_names, dims, indexes, _ = merge_data_and_coords(
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/core/dataset.py", line 426, in merge_data_and_coords
>     return merge_core(
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/core/merge.py", line 724, in merge_core
>     dims = calculate_dimensions(variables)
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/core/variable.py", line 3007, in calculate_dimensions
>     raise ValueError(
> ValueError: dimension 'time' already exists as a scalar variable
> ```
> 
>     * _Not 100% what is happening here when you open up the dataset._
> 
> 
> ### Issue 7
> 
> ```python
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/derivations/utils.py", line 175, in cosp_bin_sum
>     prs: FileAxis = cld.getAxis(0)
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/core/common.py", line 278, in __getattr__
>     raise AttributeError(
> AttributeError: 'DataArray' object has no attribute 'getAxis'
> ```
> 
>     * The `cosp_bin_sum()` and `cosp_histogram_standardize()` functions are refactored to use Xarray/xCDAT in this PR: https://github.com/E3SM-Project/e3sm_diags/pull/748/files#diff-2fd1bd98452189f43f530975ba12afc289337494f779cb5a2f97144d30820f22
> 
>     * We need to merge this PR first before this issue can be fixed.
> 
> 
> ### Issue 8
> 
> ```python
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/derivations/formulas.py", line 81, in pminuse_convert_units
>     var.attrs["units"] == "kg/m2/s"
> KeyError: 'units'
> ```
> 
>     * The `units` attribute looks like it is dropped at some point. Maybe when calculating zonal mean?

Above are remaining issues. I think they are all data/variable related, including special dimensions, calendar and units: i.e. COSP variables and PminusE. Some could be fixed with @tomvothecoder 's PR. And I think I will revisit after merging.

chengzhuzhang commented on e3sm_diags/driver/utils/io.py:
```diff
@@ -66,9 +66,7 @@ def _save_data_metrics_and_plots(
-    parameter.viewer_descr[var_key] = ds_test[var_key].attrs.get(
-        "long_name", "No long_name attr in test data"
-    )
+    parameter.viewer_descr[var_key] = ds_test[var_key].attrs.get("long_name", var_key)
```
Using shorter `var_key` may make more sense here

chengzhuzhang commented:
@tomvothecoder I think most codes for the feature is completed. I do realize new unit tests are needed. Could you please have a short demo/guide in the next Sprint? Thank you.

tomvothecoder commented:
> @tomvothecoder I think most codes for the feature is completed. Link to the viewer:[portal.nersc.gov/cfs/e3sm/chengzhu/test_e3sm_refactor/zonal_mean_xy_model_to_obs_ANN/viewer](https://portal.nersc.gov/cfs/e3sm/chengzhu/test_e3sm_refactor/zonal_mean_xy_model_to_obs_ANN/viewer/)

Awesome, thanks the update Jill.

>  I do realize new unit tests are needed. Could you please have a short demo/guide in the next Sprint? Thank you.

I believe all of the unit tests I've written are for the refactored utility modules (Dataset class, metrics functions, regridding functions, etc.). I don't have unit tests for the direct modules of the `lat_lon_driver.py` and `lat_lon_plot.py` yet (e.g., `lat_lon_driver.run_diags()`) since it will take a lot of time/work and we need to move fast. I'll think about how we can write unit tests for these complex modules in order to fit in time for it while we refactor.

chengzhuzhang commented:
when reviewing the results, I found an issue that the units for 3d variables, and the title were not showing up correctly:
![image](https://github.com/E3SM-Project/e3sm_diags/assets/13056557/83293586-1ffb-4941-a648-0e51eb502862)

tomvothecoder commented on e3sm_diags/driver/utils/dataset_xr.py:
```diff
@@ -382,7 +384,7 @@ def _get_climo_dataset(self, season: str) -> xr.Dataset:
             using other datasets.
         """
         filepath = self._get_climo_filepath(season)
-        ds = xr.open_dataset(filepath, use_cftime=True)
```
I think the tests are failing now in the CI/CD build because we always add "X" and "Y" bounds if they are missing.

The [test results](https://github.com/E3SM-Project/e3sm_diags/actions/runs/7590599128/job/20677384186) indicate bounds are missing for the expected objects. If you're up for it, this is a good opportunity for you to try to fix the 11 failing unit tests to get some experience with pytest.

You have a few options to fix the tests:
1. Update the input xr.Dataset objects to have `lat_bnds` and `lon_bnds`
   - [Example](https://github.com/E3SM-Project/e3sm_diags/blob/f171083984beebb6fd41214458ddae82fdba6caf/tests/e3sm_diags/driver/utils/test_dataset_xr.py#L218-L345)
2. Update the expected `xr.Dataset` objects to have `lat_bnds` and `lon_bnds` 
   - [Example](https://github.com/E3SM-Project/e3sm_diags/blob/f171083984beebb6fd41214458ddae82fdba6caf/tests/e3sm_diags/driver/utils/test_dataset_xr.py#L367)

tomvothecoder commented on e3sm_diags/driver/utils/dataset_xr.py:
```diff
@@ -382,7 +384,7 @@ def _get_climo_dataset(self, season: str) -> xr.Dataset:
         """
         filepath = self._get_climo_filepath(season)
-        ds = xr.open_dataset(filepath, use_cftime=True)
+        ds = xc.open_dataset(filepath, use_cftime=True, add_bounds=["X", "Y"])
```
I think the tests are failing now in the CI/CD build because we now always add "X" and "Y" bounds if they are missing through xCDAT.

The [test results](https://github.com/E3SM-Project/e3sm_diags/actions/runs/7590599128/job/20677384186) indicate bounds are missing for the expected objects. If you're up for it, this is a good opportunity for you to try to fix the 11 failing unit tests to get some experience with pytest.

You have a few options to fix the tests:
1. Update the input xr.Dataset objects to have `lat_bnds` and `lon_bnds` -- avoid adding `lat_bnds` and `lon_bnds` to the dataset when opening them with xCDAT
   - [Example](https://github.com/E3SM-Project/e3sm_diags/blob/f171083984beebb6fd41214458ddae82fdba6caf/tests/e3sm_diags/driver/utils/test_dataset_xr.py#L218-L345)
2. Update the expected `xr.Dataset` objects to have `lat_bnds` and `lon_bnds` -- takes into account newly added `lat_bnds` and `lon_bnds` generated by xCDAT
   - [Example](https://github.com/E3SM-Project/e3sm_diags/blob/f171083984beebb6fd41214458ddae82fdba6caf/tests/e3sm_diags/driver/utils/test_dataset_xr.py#L367)

chengzhuzhang replied:
Thanks for the tips! I think I should have fixed the tests. It was a good exercise with pytest...

chengzhuzhang commented on tests/e3sm_diags/driver/utils/test_dataset_xr.py:
```diff
@@ -364,7 +368,9 @@ def test_returns_reference_climo_dataset_from_file(self):
 
         ds = Dataset(parameter, data_type="ref")
         result = ds.get_ref_climo_dataset("ts", "ANN", self.ds_climo.copy())
+        print(result)
```
I should remove the two print statements.

tomvothecoder commented:
> ### Issue 6
> ```python
>  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/e3sm_diags/driver/utils/dataset_xr.py", line 254, in _get_global_attr_from_climo_dataset
>     ds = xr.open_dataset(filepath)
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/backends/api.py", line 572, in open_dataset
>     backend_ds = backend.open_dataset(
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/backends/netCDF4_.py", line 607, in open_dataset
>     ds = store_entrypoint.open_dataset(
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/backends/store.py", line 58, in open_dataset
>     ds = Dataset(vars, attrs=attrs)
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/core/dataset.py", line 697, in __init__
>     variables, coord_names, dims, indexes, _ = merge_data_and_coords(
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/core/dataset.py", line 426, in merge_data_and_coords
>     return merge_core(
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/core/merge.py", line 724, in merge_core
>     dims = calculate_dimensions(variables)
>   File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/e3sm_diags_dev_654_zonal_mean_xy/lib/python3.10/site-packages/xarray/core/variable.py", line 3007, in calculate_dimensions
>     raise ValueError(
> ValueError: dimension 'time' already exists as a scalar variable
> ```
> * _Not 100% what is happening here when you open up the dataset._

I am running into this issue with `dataset_xr.Dataset._get_global_attr_from_climo_dataset()` in #748. The issue is the dataset incorrectly includes `"time"` as a scalar variable rather than as coordinates. In this case, Xarray is correct to raise an error when trying to open the dataset (below quote).

> This should actually raise an error at construction time for the DataArray. Variables with names matching dimensions are special in xarray: they are required to be 1D arrays with length matching the dimension size.
> - https://github.com/pydata/xarray/issues/1709

Example Code:
```python
import xarray as xr
import cdms2

filepath = "/global/cfs/cdirs/e3sm/diagnostics/observations/Atm/climatology/MISRCOSP/MISRCOSP_ANN_climo.nc"

# ValueError: dimension 'time' already exists as a scalar variable
ds1 = xr.open_dataset(filepath)

ds2 = cdms2.open(filepath)
print(ds2.variables)

"""
{'gw': <cdms2.fvariable.FileVariable at 0x7fad9618f7c0>,
 'CLDHGH_MISR': <cdms2.fvariable.FileVariable at 0x7fad95c8fa00>,
 'CLDINT_MISR': <cdms2.fvariable.FileVariable at 0x7fad95c8e2c0>,
 'CLDLOW_MISR': <cdms2.fvariable.FileVariable at 0x7fad95c8f7f0>,
 'CLDMED_MISR': <cdms2.fvariable.FileVariable at 0x7fad9602fa60>,
 'CLDTHICK_MISR': <cdms2.fvariable.FileVariable at 0x7fad9602edd0>,
 'CLDTHN_MISR': <cdms2.fvariable.FileVariable at 0x7fb1d94ee050>,
 'CLDTOT_MISR': <cdms2.fvariable.FileVariable at 0x7fb1d94ed930>,
 'CLMISR': <cdms2.fvariable.FileVariable at 0x7fb1d94edde0>,
 'misr_cth_bnds': <cdms2.fvariable.FileVariable at 0x7fb1d94ee2c0>,
 'misr_tau_bnds': <cdms2.fvariable.FileVariable at 0x7fb1d94edf30>,
 'time': <cdms2.fvariable.FileVariable at 0x7fb1d94ee0e0>}
"""
```

I'll check if writing a `preprocess()` function to convert `"time"` to coordinates is a possible workaround to this dataset issue ([source](https://github.com/pydata/xarray/discussions/8197#discussioncomment-7027224)).

tomvothecoder commented:
> I'll check if writing a `preprocess()` function to convert `"time"` to coordinates is a possible workaround to this dataset issue ([source](https://github.com/pydata/xarray/discussions/8197#discussioncomment-7027224)).

This doesn't work because it still opens the dataset first then applies the function to it. Any thoughts on how to deal with this dataset issue?

chengzhuzhang commented:
> This doesn't work because it still opens the dataset first then applies the function to it.
> 
> `drop_variables` looks like it works. Is the "time" scalar variable ever used

Good to know there is a workaround. I don't think the "time" scalar variable is ever used for this dataset. Though I don't have a good idea how to implement this dataset specific solution.

tomvothecoder commented:
> > This doesn't work because it still opens the dataset first then applies the function to it.
> > `drop_variables` looks like it works. Is the "time" scalar variable ever used
> 
> Good to know there is a workaround. I don't think the "time" scalar variable is ever used for this dataset. Though I don't have a good idea how to implement this dataset specific solution.
> 
> We can perhaps fix it in the source data?

I just implemented a workaround in PR #748 [here](https://github.com/E3SM-Project/e3sm_diags/pull/748/commits/d57d6ac1ce5db7c501affb809a70203c7a27d2e9#diff-35006a05805f6f5876c9fe70b5281c865955b8a82d1574fd7f62d6efda623975). Basically, if this specific issue ever pops up for a dataset then the "time" scalar variable is dropped and a new "time" scalar coordinate is added. 

```python
    def _open_climo_dataset(self, filepath: str) -> xr.Dataset:
        """Open a climatology dataset.

        Some climatology files have "time" as a scalar variable. If this
        variable is not a 1D array with a length matching the equivalent
        dimension size, Xarray will `raise ValueError: dimension 'time'
        already exists as a scalar variable`. We drop the "time" variable
        as a workaround to this issue for these cases and add new "time"
        coordinates.

        Parameters
        ----------
        filepath : str
            The path to a climatology dataset.

        Returns
        -------
        xr.Dataset
            The climatology dataset.

        Raises
        ------
        ValueError
            Raised for all ValueErrors other than "dimension 'time' already
            exists as a scalar variable".
        """
        args = {"path": filepath, "use_cftime": True, "add_bounds": ["X", "Y"]}

        try:
            ds = xc.open_dataset(**args)
        except ValueError as e:
            msg = str(e)

            if "dimension 'time' already exists as a scalar variable" in msg:
                ds = xc.open_dataset(**args, drop_variables=["time"])
                ds.coords["time"] = xr.DataArray(
                    name="time",
                    dims=["time"],
                    data=[0],
                    attrs={"axis": "T", "standard_name": "time"},
                )
            else:
                raise ValueError(msg)

        return ds
```

chengzhuzhang commented:
This is a good solution!

chengzhuzhang commented on e3sm_diags/derivations/formulas.py:
```diff
@@ -77,15 +77,16 @@ def qflx_convert_to_lhflx_approxi(var: xr.DataArray):
-        # var = convert_units( var, 'kg/m2/s' )
-        var = var * 3600.0 * 24  # convert to mm/day
-        var.attrs["units"] = "mm/day"
+    if hasattr(var, "units"):
```
Fixing issue 8 in https://github.com/E3SM-Project/e3sm_diags/pull/752#issuecomment-1901381465. xarray flag an error when dataarray doesn't have a units, while `cdms2` doesn't.

chengzhuzhang replied:
I mean `cdms2` doesn't flag an error in this case.

chengzhuzhang commented on e3sm_diags/driver/utils/dataset_xr.py:
```diff
@@ -386,10 +386,10 @@ def _get_climo_dataset(self, season: str) -> xr.Dataset:
-        elif self.var in self.derived_vars_map:
+        if self.var in self.derived_vars_map:
             ds = self._get_dataset_with_derived_climo_var(ds)
+        elif self.var in ds.variables:
```
Revert back the logic of derived variable as codes on main. This fix a problem that variables are not converted, for situations when self.var in ds.variables, but they still need units conversion or renaming.

chengzhuzhang commented on e3sm_diags/driver/utils/regrid.py:
```diff
@@ -526,7 +526,8 @@ def _hybrid_to_plevs(
         )
-
+    # Keep units unchanged.
+    result[var_key].attrs["units"] = ds[var_key].units
```
somehow the units are not preserved with vertical regridding, not sure if its `xgcm` or  and `xcdat` problem.

tomvothecoder replied:
Based on this xcdat [code](https://github.com/xCDAT/xcdat/blob/fbf1db68fd6328d5a534e0c1e57e22e6fdd70c76/xcdat/regridder/xgcm.py#L200-L208), `xcdat` _should_ be preserving attributes (while `xgcm` does not), 

I added `xr.set_options(keep_attrs=True)` to this code block just to make sure. The unit tests for `regrid_z_axis_to_plevs()` also check to make sure attributes are preserved/updated ([example](https://github.com/E3SM-Project/e3sm_diags/blob/1530283382ac2012c77d820a2f5d24d0df5e7618/tests/e3sm_diags/driver/utils/test_regrid.py#L329-L353)). 

I'm not 100% sure if this line is needed here. I'll step through the code to check.

tomvothecoder replied:
I stepped through the code and found that the units _are_ correctly being preserved in xCDAT. Since the variable is being regridded from hybrid to pressure levels, the units of the pressure level is being used ("mb").

Are we supported to maintain the original units of the variable? The old CDAT version of this function also sets "mb" as the variable's units:

https://github.com/E3SM-Project/e3sm_diags/blob/a6bbc1c893ff81c480cc0793409f4889f7f76d5d/e3sm_diags/driver/utils/general.py#L131-L137

chengzhuzhang replied:
Yes, we should maintain the original units of the variable. The CDAT version of the function is correct, it sets `mb` (milibar) as the units of the reconstructed 3d pressure. The corresponding refactored code is here https://github.com/E3SM-Project/e3sm_diags/blob/b1183a7375f058695d842814bef2c55ace55b480/e3sm_diags/driver/utils/regrid.py#L534-L586: 
The refactored code correctly set `mb` for pressure.

The units of the variable should be preserved here and also in the next vertical interpolation operation.

tomvothecoder replied:
Got it. We can keep your changes here and I will mark this thread as a resolved.

chengzhuzhang commented:
Ready for review. Issues 4, 5, 6, 7 remain, but 6 and 7 are expected to be fixed in @tomvothecoder 's cosp histogram PR.

tomvothecoder commented:
@chengzhuzhang I am close to done with PR #748. Let's merge that PR first, then get those fixes into your branch for this PR. 

I'm also working on another regression testing notebook that compares `.nc` files. The existing one only compares `.json` files, which some sets do not produce. We can adapt that notebook for this branch for testing.

chengzhuzhang commented:
@tomvothecoder Thanks for the heads up. Your plan sounds good!

tomvothecoder reviewed (approved):
Here is my initial review with some minor updates that I will commit directly. The PR should be good to go after regression testing and fixing the integration test bug.


I also pushed some changes directly in this commit [`b19aec0` (#752)](https://github.com/E3SM-Project/e3sm_diags/pull/752/commits/b19aec0c166423177772f3291f873a6794dcb71d) with these changes:
- Update instructions for `template_run_script.py` to run the script as a Python module
- Add `panel_configs` and `border_padding` args to `plot.utils.save_plot()`
- Delete `_save_plot()` in `zonal_mean_xy_plot.py` and replace with `plot.utils.save_plot()`
- Update `_calc_zonal_mean()` to reuse `metrics.spatial_avg()`

tomvothecoder commented on auxiliary_tools/cdat_regression_testing/template_run_script.py:
```diff
@@ -12,11 +12,13 @@
+6. Run this script as a Python module
+   - `auxiliary_tools` is not included in `setup.py`, so `-m` is required
+     to run the script as a Python module
+   - Command: python -m auxiliary_tools.cdat_regression_testing.<ISSUE-<SET_NAME>/<SCRIPT-NAME>
```
```suggestion
   - Command: python -m auxiliary_tools.cdat_regression_testing.<ISSUE-<SET_NAME./<SCRIPT-NAME>
```

tomvothecoder commented on e3sm_diags/plot/utils.py:
```diff
@@ -52,6 +57,12 @@ def _save_plot(fig: plt.Figure, parameter: CoreParameter):
     parameter : CoreParameter
         The CoreParameter with file configurations.
+    panel_configs : List[Tuple[float, float, float, float]]
+        A list of panel configs consiting of positions and sizes, with each
```
```suggestion
        A list of panel configs consisting of positions and sizes, with each
```

chengzhuzhang commented:
Based on the[ regression test result](https://github.com/E3SM-Project/e3sm_diags/blob/6c439b665ff28e05f01c0c7d87a2e9a47ce98f08/auxiliary_tools/cdat_regression_testing/654-zonal_mean_xy/654-zonal_mean_xy_cdat_regression_test_netcdf.ipynb), several problems are found. They are grouped as follows:
1. Variables not generated and saved for `NETCF`, `RESTOM` and `FLDSC`, and all simulator related variables: e.g. `CLDTOT_TAU1.3_ISCCP`. error messsage:
`OSError: Variable 'NETCF' was not in the file '/global/cfs/cdirs/e3sm/e3sm_diags/postprocessed_e3sm_v2_data_for_e3sm_diags/20210528.v2rc3e.piControl.ne30pg2_EC30to60E2r2.chrysalis/climatology/rgr/20210528.v2rc3e.piControl.ne30pg2_EC30to60E2r2.chrysalis_ANN_005101_006012_climo.nc', nor was it defined in the derived variables dictionary.`
2. Large difference between saved `CERES-EBAF-TOA-v4.1` reference data comparison. 
3. Shape mismatch for subset of ERA5, and MERRA2 reference data comparison.
TMQ e.g. `(shapes (180,), (361,) mismatch)`

tomvothecoder commented on e3sm_diags/driver/utils/dataset_xr.py:
```diff
@@ -385,17 +387,22 @@ def _get_climo_dataset(self, season: str) -> xr.Dataset:
-        if self.var in ds.variables:
-            pass
-        elif self.var in self.derived_vars_map:
+        if self.var in self.derived_vars_map and (self.var,) in list(
+            self.derived_vars_map[self.var].keys()
+        ):
             ds = self._get_dataset_with_derived_climo_var(ds)
+        elif self.var in ds.variables:
+            pass
```
```suggestion
        if self.var in self.derived_vars_map:
            ds = self._get_dataset_with_derived_climo_var(ds)
        elif self.var in ds.variables:
            pass
```

I think we just need to switch up the order of the conditionals to what it was in the old [Dataset._get_climo_var()](https://github.com/E3SM-Project/e3sm_diags/blob/10b16f8a52d5dc606c9060de7b517c01ca4d804b/e3sm_diags/driver/utils/dataset.py#L346-L372)

Any derived variable with a unit conversion formula should run that function.

tomvothecoder replied:
This change might also fix issue #1 from your [comment above](https://github.com/E3SM-Project/e3sm_diags/pull/752#issuecomment-1927819504):

> 1. Variables not generated and saved for `NETCF`, `RESTOM` and `FLDSC`, and all simulator related variables: e.g. `CLDTOT_TAU1.3_ISCCP`. error messsage:
>        `OSError: Variable 'NETCF' was not in the file '/global/cfs/cdirs/e3sm/e3sm_diags/postprocessed_e3sm_v2_data_for_e3sm_diags/20210528.v2rc3e.piControl.ne30pg2_EC30to60E2r2.chrysalis/climatology/rgr/20210528.v2rc3e.piControl.ne30pg2_EC30to60E2r2.chrysalis_ANN_005101_006012_climo.nc', nor was it defined in the derived variables dictionary.`

tomvothecoder replied:
In my PR #774, this update resolves the large relative diffs I was getting for OMEGA and Q variables which require unit conversion for some cases.

chengzhuzhang replied:
@tomvothecoder thanks for looking into this issue. with your suggested change, it resolved the first of my 3 remaining regression problems. Though a new CI/CD problem occur, I think it is with the logic of derived variable changed back (shifted order of conditionals)...

chengzhuzhang commented:
For the 3rd problem in regression test: 
It turned out when the reference data is not available or processed. (it (`ds_ref` will be the same dataset as `ds_test`)
 https://vscode.dev/github/E3SM-Project/e3sm_diags/blob/refactor/654-zonal_mean_xy/e3sm_diags/driver/zonal_mean_xy_driver.py#L136. This caused a dimension mismatch. We need to fix this assumption and allow only test data is being processed/saved and plotted.

tomvothecoder commented on e3sm_diags/driver/utils/dataset_xr.py:
```diff
@@ -385,17 +387,22 @@ def _get_climo_dataset(self, season: str) -> xr.Dataset:
-        if self.var in ds.variables:
-            pass
-        elif self.var in self.derived_vars_map:
+        if self.var in self.derived_vars_map and (self.var,) in list(
+            self.derived_vars_map[self.var].keys()
+        ):
             ds = self._get_dataset_with_derived_climo_var(ds)
```
```suggestion
	if self.var in self.derived_vars_map:
```

I noticed you pushed the above change in [`daa748e` (#752)](https://github.com/E3SM-Project/e3sm_diags/pull/752/commits/daa748e06953b34db1fb76c8a67f5df1446183d6) but reverted it in [`7a494ef` (#752)](https://github.com/E3SM-Project/e3sm_diags/pull/752/commits/7a494ef514062981ad38ba4ad41ef3b365c486f8#diff-35006a05805f6f5876c9fe70b5281c865955b8a82d1574fd7f62d6efda623975).

Was this on accident? This explains why the CI/CD build is still breaking.

Also are you certain you are still running into issues 3 and 4 in this [comment](https://github.com/E3SM-Project/e3sm_diags/pull/752#issuecomment-1927819504) with this update? A re-test might be a good idea.

tomvothecoder commented on e3sm_diags/driver/utils/regrid.py:
```diff
@@ -526,7 +526,8 @@ def _hybrid_to_plevs(
         )
-
+    # Keep units unchanged.
+    result[var_key].attrs["units"] = ds[var_key].units
```
I'm copied this fix over to PR #774 in [`ecc2102` (#774)](https://github.com/E3SM-Project/e3sm_diags/pull/774/commits/ecc2102d75213809a24eae5967323bdbf10b6b56) since that PR will be merged shortly.

chengzhuzhang commented:
I followed the suggested change pointed out by https://github.com/E3SM-Project/e3sm_diags/pull/752#discussion_r1490045128. It fixed the CI/CD, but with remaining issues from regression tests, summarized here. 

1. A different set of 21 variables are not created: `FLUT`, `FLUTC`, `SOLIN`, `FLDS`, `FLNS`, `FLNSC`, `FSDS`, `FLDSC`, `FSNS`, `FSNSC`,`ISCCPCOSP-CLDTOT_TAU1.3_9.4_ISCCP`, `ISCCPCOSP-CLDTOT_TAU1.3_ISCCP`, `MODISCOSP-CLDHGH/TOT_TAU1.3_9.4_MODIS`, `MODISCOSP-CLDHGH/TOT_TAU1.3_MODIS`, `ERA5-TMQ`, `MERRA2-TMQ`

2. Variables diff between refactored vs main: 
[645_unmatched.txt](https://github.com/E3SM-Project/e3sm_diags/files/14289669/645_unmatched.txt)

@tomvothecoder I'm documenting the remaining issues. I think these will recur in lat_lon and polar testing. We can merge this branch for now and try to fix in another PR.

tomvothecoder commented:
> I followed the suggested change pointed out by [#752 (comment)](https://github.com/E3SM-Project/e3sm_diags/pull/752#discussion_r1490045128). It fixed the CI/CD, but with remaining issues from regression tests, summarized here.
> 
>     1. A different set of 21 variables are not created: `FLUT`, `FLUTC`, `SOLIN`, `FLDS`, `FLNS`, `FLNSC`, `FSDS`, `FLDSC`, `FSNS`, `FSNSC`,`ISCCPCOSP-CLDTOT_TAU1.3_9.4_ISCCP`, `ISCCPCOSP-CLDTOT_TAU1.3_ISCCP`, `MODISCOSP-CLDHGH/TOT_TAU1.3_9.4_MODIS`, `MODISCOSP-CLDHGH/TOT_TAU1.3_MODIS`, `ERA5-TMQ`, `MERRA2-TMQ`
> 
>     2. Variables diff between refactored vs main:
>        [645_unmatched.txt](https://github.com/E3SM-Project/e3sm_diags/files/14289669/645_unmatched.txt)
> 
> 
> @tomvothecoder I'm documenting the remaining issues. I think these will recur in lat_lon and polar testing. We can merge this branch for now and try to fix in another PR. Could you please help merging?

Got it, we can transfer these to another GitHub issue.

tomvothecoder commented:
Actually, I'll take a brief look at these issues and try to debug while I'm at it.

chengzhuzhang commented:
> Actually, I'll take a brief look at these issues and try to debug while I'm at it.

Thank you!

tomvothecoder commented on e3sm_diags/driver/utils/dataset_xr.py:
```diff
@@ -627,32 +639,39 @@ def _get_dataset_with_derived_climo_var(self, ds: xr.Dataset) -> xr.Dataset:
+            # 1. Get the derivation function.
+            derivation_func = list(matching_target_var_map.values())[0]
 
-        # 2. Get the derivation function arguments using source variable keys.
-        # Example: [xr.DataArray(name="PRECC",...), xr.DataArray(name="PRECL",...)]
-        src_var_keys = list(matching_target_var_map.keys())[0]
+            # 2. Get the derivation function arguments using source variable keys.
+            # Example: [xr.DataArray(name="PRECC",...), xr.DataArray(name="PRECL",...)]
+            src_var_keys = list(matching_target_var_map.keys())[0]
 
-        # 3. Use the derivation function to derive the variable.
-        ds_final = self._get_dataset_with_derivation_func(
-            ds, derivation_func, src_var_keys, target_var
-        )
+            # 3. Use the derivation function to derive the variable.
+            ds_derived = self._get_dataset_with_derivation_func(
+                ds, derivation_func, src_var_keys, target_var
+            )
 
-        return ds_final
+            return ds_derived
+
+        # None of the entries in the derived variables dictionary worked,
+        # so try to get the variable directly from he dataset.
+        if target_var in ds.data_vars.keys():
+            return ds
+
+        raise IOError(
+            f"The dataset file has no matching source variables for {target_var}"
+        )
```
RE: Issue 1 in https://github.com/E3SM-Project/e3sm_diags/pull/752#issuecomment-1945057995

This fixes the missing outputs for the following variables: `FLUT`, `FLUTC`, `SOLIN`, `FLDS`, `FLNS`, `FLNSC`, `FSDS`, `FLDSC`, `FSNS`, `FSNSC`, `ERA5-TMQ`, `MERRA2-TMQ`

The issue was that I did not include a conditional to try to return the variable directly from the climatology dataset if it cannot be derived (lines 661-664)

tomvothecoder commented on e3sm_diags/derivations/formulas_cosp.py:
```diff
@@ -116,19 +116,20 @@ def cosp_bin_sum(target_var_key: str, var: xr.DataArray) -> xr.DataArray:
 
-    # 4. Get the axes mask conditional and subset the variable on it.
+    # 5. Get the axes mask conditional and subset the variable on it.
     cond = (
         (prs >= prs_range[0])
         & (prs <= prs_range[-1])
         & (tau >= tau_range[0])
         & (tau <= tau_range[-1])
     )
+
     var_sub = var.where(cond, drop=True)
 
-    # 5. Sum on axis=0 and axis=1 (tau and prs)
+    # 7. Sum on axis=0 and axis=1 (tau and prs)
     var_sum = var_sub.sum(dim=[prs.name, tau.name])
```
RE: Issue 1 in https://github.com/E3SM-Project/e3sm_diags/pull/752#issuecomment-1945057995

This fixes missing outputs for the remaining variables: `ISCCPCOSP-CLDTOT_TAU1.3_9.4_ISCCP`, `ISCCPCOSP-CLDTOT_TAU1.3_ISCCP`, `MODISCOSP-CLDHGH/TOT_TAU1.3_9.4_MODIS`, `MODISCOSP-CLDHGH/TOT_TAU1.3_MODIS`

### Root Cause

1. The `prs_range` can be in descending order (e.g., `(90000.0, 9000.0)`) 
2. This caused the condition (`cond`) to not match any of the data values when subsetting with `.where()`
   - The condition being used is `>=90000.0 and <=9000.0` which is incorrect
   - The _correct_ condition should to be the other way around, `>= 9000 and <= 90000`
 3. Since there are no matching values, the below `numpy` error is thrown with `drop=True` (drop all `nan` values that don't match): 
   - `ValueError: Cannot apply_along_axis when any iteration dimensions are 0`

### Solution 

Sort `prs_range` and `tau_range` before subsetting

### Side-note

The CDMS2 version of sub-setting a variable is pretty confusing. Notice below how `prs_low` and `prs_high` are used to subset, but `prs_low` is actually the larger value. 

I think CDMS2 swaps both values around and subsets to include all values within the range. It uses the correct sub-setting condition, `>= 9000 and <= 90000`. 

https://github.com/E3SM-Project/e3sm_diags/blob/434ab8cdd0aa0781e0064dafb0309695e714960d/e3sm_diags/derivations/acme.py#L488-L490

tomvothecoder commented on e3sm_diags/metrics/metrics.py:
```diff
@@ -21,33 +21,39 @@ def spatial_avg(
-    Notes
-    -----
-    Replaces `e3sm_diags.metrics.mean`.
+        Parameters
+        ----------
+        ds : xr.Dataset
+            The dataset containing the variable.
+        var_key : str
+    <<<<<<< HEAD
+            The key of the variable in the dataset.
+        axis : List[Axis]
+            The list of axes to use for the computation, by default ["X", "Y"].
+            Valid axes including "X", "Y", and "Z".
+    =======
+            The key of the variable.
+        axis : List[str]
+            A list of axis strings, by default ["X", "Y"].
+    >>>>>>> b7195dad (Refactor zonal_mean_xy set)
+        as_list : bool
+            Return the spatial average as a list of floats, by default True.
+            If False, return an xr.DataArray. Must be True to be serializable for
+            writing out to a `.json` metrics file.
```
```suggestion
        Parameters
        ----------
        ds : xr.Dataset
            The dataset containing the variable.
        var_key : str
            The key of the variable in the dataset.
        axis : List[Axis]
            The list of axes to use for the computation, by default ["X", "Y"].
            Valid axes including "X", "Y", and "Z".
            The key of the variable.
        axis : List[str]
            A list of axis strings, by default ["X", "Y"].
        as_list : bool
            Return the spatial average as a list of floats, by default True.
            If False, return an xr.DataArray. Must be True to be serializable for
            writing out to a `.json` metrics file.
```

tomvothecoder reviewed (approved):
I resolved issue #1 your comment [here](https://github.com/E3SM-Project/e3sm_diags/pull/752#issuecomment-1945057995) with two separate fixes (mentioned in my review comments)


Here's my debugging [notebook](https://github.com/E3SM-Project/e3sm_diags/pull/752/files#diff-cb918f09702a9bde1a520ddbc8bd03a9c8121812082bb541ac76595a31d477b5) now showing `Number of files missing: 0`

I will merge this PR once the build passes.

chengzhuzhang commented:
@tomvothecoder Thank you for the perseverance to fix the derived variable logic and those COSP variable calculation!!
